### ISSUE TITLE
Allow globalState to be synchronized across multiple windows

### ISSUE THREAD

**Issue by fabiospampinato** [Microsoft/vscode#55834]
- VSCode Version: 1.25.1
- OS Version: 10.13.1

I'm trying to use `ExtensionContext.globalState` in my [GitHub Notifications Bell](https://marketplace.visualstudio.com/items?itemName=fabiospampinato.vscode-github-notifications-bell) extension, which notifies the user about GitHub notifications, in order to avoid querying GitHub's API multiple times when there are multiple windows open, basically I want to share some state between all the instances of the extension.

But I think the way `globalState` is implemented is buggy. I have the following piece of code, which is called every 10 seconds:

```typescript
  // Utils.state === context.globalState => true
  async function updateState () {
    const now = Date.now ();
    if ( ( now - Utils.state.get ( 'date', 0 ) ) >= 30000 ) { // Update, at least 30s have passed
      vscode.window.showInformationMessage ( `P ${Utils.state.get ( 'date', 0 )} N ${now} D ${now - Utils.state.get ( 'date', 0 )}` );
      await Utils.state.update ( 'date', now );
      // Refresh logic...
    } else {
      vscode.window.showInformationMessage ( `Skipping ${Utils.state.get ( 'date', 0 )}` );
    }
  }
```

It is supposed to update the state only if the last update happened at least 30s ago. I see this happening:

![Bug](https://user-images.githubusercontent.com/1812093/43686181-e3db5c4a-98c1-11e8-87c9-38c7ee6a625b.gif)

1. At first `Utils.state.get ( 'date', 0 )` returns the same number in each window
2. **All** windows update the state (this should only happen when there's a race condition, which is not what's happening here unless `Utils.state.update ( 'date', now )` is taking forever)
3. Now each window gets a different value from `Utils.state.get ( 'date', 0 )`, basically the state is not global anymore

**Comment by bpasero** [comment]
@fabiospampinato so you say this works when you use it in the same window but not when using across multiple windows?

**Comment by bpasero** [comment]
@fabiospampinato is that method you provided reproducing the issue easily? I would like to reproduce with a simple hello world extension if you could provide that code.

**Comment by fabiospampinato** [comment]
Yes, I can reproduce with the following extension:

```ts

import * as vscode from 'vscode';

function activate ( context: vscode.ExtensionContext ) {

  async function updateState () {
    const now = Date.now ();
    if ( ( now - context.globalState.get ( 'date', 0 ) ) >= 30000 ) { // Update, at least 30s have passed
      vscode.window.showInformationMessage ( `P ${context.globalState.get ( 'date', 0 )} N ${now} D ${now - context.globalState.get ( 'date', 0 )}` );
      await context.globalState.update ( 'date', now );
      // Refresh logic...
    } else {
      vscode.window.showInformationMessage ( `Skipping ${context.globalState.get ( 'date', 0 )}` );
    }
  }

  setInterval ( updateState, 10000 );

}

export {activate};
```

Just make sure the extension gets activated, then you'll see something analogous to what I've recorded in the gif. 

Here's the extension:

[GlobalStateBug-1.2.0.vsix.zip](https://github.com/Microsoft/vscode/files/2282790/GlobalStateBug-1.2.0.vsix.zip)

**Comment by bpasero** [comment]
@jrieken when I look into the implementation of how the `localStorage` contents are surfaced to extensions (in [`ExtensionMemento`](https://github.com/Microsoft/vscode/blob/bbd54cb673382d74dbc252c2ffb85e5eadcc4bd8/src/vs/workbench/api/node/extHostExtensionService.ts#L40)) it looks like the value is retrieved once and then used to answer any further requests instead of going back to the main side. That means, if another window changes that (gobal) extension state, the original window will not see this change.

In other words it does not look like we support global shared state between an extension running in multiple windows via `localStorage` at least.

**Comment by jrieken** [comment]
> That means, if another window changes that (gobal) extension state, the original window will not see this change.

Yeah, good find. That's something we got wrong since day one. As usual, reading is sync, writing async. What we need is an event if the storage as been changed and then update the shared, global state

**Comment by bpasero** [comment]
Yeah, such an event does not exist currently but maybe can be added when we move to a different storage mechanism.

**Comment by fabiospampinato** [comment]
@ShyykoSerhiy in my particular use case that can be used for avoiding over-querying the API, but then  some windows will display the icon even when not necessary, which kinds of the defeats the purpose of my extension.

**Comment by sandy081** [comment]
@bpasero Does this synchronisation happens even on main side? And can we have events to storage changes which is needed to fix #57422

**Comment by bpasero** [comment]
@sandy081 today `localStorage` can only be accessed from a window, never main side.

**Comment by sandy081** [comment]
Sorry I mean window(renderer) side.

**Comment by bpasero** [comment]
@sandy081 this would only work for global state right? You want a change event in the one window when it changes in the other window?

**Comment by sandy081** [comment]
Yes, we store visibility state globally and when a window changes global state then other window should get a change event.

**Comment by bpasero** [comment]
@sandy081 yeah I think the new solution should cover that. let me know if there is something else we should cover when working on a re-implementation besides that.

**Comment by bpasero** [comment]
**Verification:** write an extension that adds data into `context.globalState` and run this extension from multiple windows. Verify that each window always has the latest data that was stored last.
